File: middlewarehouse/models/query_statement_test.go
package models

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

type QueryStatementTestSuite struct {
	suite.Suite
	assert *assert.Assertions
}

func TestQueryStatementSuite(t *testing.T) {
	suite.Run(t, new(QueryStatementTestSuite))
}

func (suite *QueryStatementTestSuite) Test_JsonDeserialization_Success() {
	jsonStatement := []byte(`
	{
		"comparison": "and",
		"conditions": [
			{
				"rootObject": "ShippingAddress",
				"field": "address1",
				"operator": "notContains",
				"value": "p.o. box"
			},
			{
				"rootObject": "ShippingAddress",
				"field": "countryId",
				"operator": "equals",
				"value": 1
			}
		]
	}`)

	statement := new(QueryStatement)
	err := json.Unmarshal(jsonStatement, statement)
	suite.Nil(err)

	suite.Equal(And, statement.Comparison)
	suite.Len(statement.Conditions, 2)

	condition1 := statement.Conditions[0]
	suite.Equal("ShippingAddress", condition1.RootObject)
	suite.Equal("address1", condition1.Field)
	suite.Equal(NotContains, condition1.Operator)
	suite.Equal("p.o. box", condition1.Value.(string))

	condition2 := statement.Conditions[1]
	suite.Equal("ShippingAddress", condition2.RootObject)
	suite.Equal("countryId", condition2.Field)
	suite.Equal(Equals, condition2.Operator)
	suite.Equal(1, int(condition2.Value.(float64)))
}

File: phoenix-scala/phoenix/test/unit/models/rules/QueryStatementTest.scala
package models.rules

import org.json4s._
import org.json4s.jackson.JsonMethods._
import phoenix.models.rules.{Condition, QueryStatement}
import phoenix.utils.JsonFormatters._
import testutils.TestBase

class QueryStatementTest extends TestBase {
  implicit val formats = phoenixFormats

  "QueryStatement" - {
    "JSON deserialization" - {
      "domestic shipping to non-P.0. boxes from JSON" in {
        val json = """
            | {
            |   "comparison": "and",
            |   "conditions": [
            |     {
            |       "rootObject": "ShippingAddress",
            |       "field": "address1",
            |       "operator": "notContains",
            |       "valString": "p.o. box"
            |     },
            |     {
            |       "rootObject": "ShippingAddress",
            |       "field": "countryId",
            |       "operator": "equals",
            |       "valInt": 1
            |     }
            |   ]
            | }
          """.stripMargin

        val statement = parse(json).extract[QueryStatement]
        statement.comparison must === (QueryStatement.And)

        val condition1 :: condition2 :: Nil = statement.conditions
        condition1.rootObject must === ("ShippingAddress")
        condition1.field must === ("address1")
        condition1.operator must === (Condition.NotContains)
        condition1.valString must === (Some("p.o. box"))

        condition2.rootObject must === ("ShippingAddress")
        condition2.field must === ("countryId")
        condition2.operator must === (Condition.Equals)
        condition2.valInt must === (Some(1))
      }

      "order subtotal is greater than $50 and less than $200" in {
        val json = """
            | {
            |   "comparison": "and",
            |   "conditions": [
            |     {
            |       "rootObject": "Order",
            |       "field": "subtotal",
            |       "operator": "greaterThanOrEquals",
            |       "valInt": 50
            |     },
            |     {
            |       "rootObject": "Order",
            |       "field": "subtotal",
            |       "operator": "lessThan",
            |       "valInt": 200
            |     }
            |   ]
            | }
          """.stripMargin

        val statement = parse(json).extract[QueryStatement]
        statement.comparison must === (QueryStatement.And)

        val condition1 :: condition2 :: Nil = statement.conditions
        condition1.rootObject must === ("Order")
        condition1.field must === ("subtotal")
        condition1.operator must === (Condition.GreaterThanOrEquals)
        condition1.valInt must === (Some(50))

        condition2.rootObject must === ("Order")
        condition2.field must === ("subtotal")
        condition2.operator must === (Condition.LessThan)
        condition2.valInt must === (Some(200))
      }
    }
  }
}
